### PR TITLE
Dynamic OpenCL batch normalization inference.

### PR DESCRIPTION
**Pull request by daniellowell**
This PR is specifically for inference. Low batch sizes do better than what we had previously, high batch sizes do worse.
Since this is for inference I suspect for now we don't have to worry a lot for larger batch sizes. 

Data collected on a MI50.

Batch size 1:

  | spatial PR | spatial dev | runtime ratio | per act PR | per act dev | runtime ratio
-- | -- | -- | -- | -- | -- | --
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 1024 -H 14 -W 14   -m ${1} --forw 2 -s 1 -r 1 | 0.010 | 0.013 | 74% | 0.015 | 0.021 | 69%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 128 -H 28 -W 28   -m ${1} --forw 2 -s 1 -r 1 | 0.011 | 0.009 | 132% | 0.008 | 0.012 | 66%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 2048 -H 7 -W 7   -m ${1} --forw 2 -s 1 -r 1 | 0.011 | 0.009 | 115% | 0.014 | 0.013 | 107%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 256 -H 14 -W 14   -m ${1} --forw 2 -s 1 -r 1 | 0.005 | 0.005 | 94% | 0.008 | 0.008 | 104%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 256 -H 56 -W 56   -m ${1} --forw 2 -s 1 -r 1 | 0.023 | 0.027 | 83% | 0.065 | 0.072 | 90%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 512 -H 28 -W 28   -m ${1} --forw 2 -s 1 -r 1 | 0.012 | 0.014 | 86% | 0.038 | 0.038 | 100%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 512 -H 7 -W 7 -m   ${1} --forw 2 -s 1 -r 1 | 0.005 | 0.005 | 100% | 0.007 | 0.006 | 110%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 64 -H 112 -W 112   -m ${1} --forw 2 -s 1 -r 1 | 0.016 | 0.025 | 63% | 0.045 | 0.072 | 62%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 1 -c 64 -H 56 -W 56   -m ${1} --forw 2 -s 1 -r 1 | 0.007 | 0.009 | 78% | 0.014 | 0.015 | 93%

Batch size 128:

  | spatial PR | spatial dev | runtime ratio | per act PR | per act dev | runtime ratio
-- | -- | -- | -- | -- | -- | --
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 128 -c 1024 -H 14 -W   14 -m ${1} --forw 2 -s 1 -r 1 | 0.347 | 0.352 | 99% | 0.314 | 0.323 | 97%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 128 -c 128 -H 28 -W   28 -m ${1} --forw 2 -s 1 -r 1 | 0.157 | 0.207 | 76% | 0.157 | 0.182 | 86%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 128 -c 2048 -H 7 -W 7   -m ${1} --forw 2 -s 1 -r 1 | 0.258 | 0.178 | 145% | 0.263 | 0.166 | 159%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 128 -c 256 -H 14 -W   14 -m ${1} --forw 2 -s 1 -r 1 | 0.103 | 0.093 | 111% | 0.104 | 0.127 | 81%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 128 -c 256 -H 56 -W   56 -m ${1} --forw 2 -s 1 -r 1 | 1.036 | 1.050 | 99% | 1.040 | 1.053 | 99%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 128 -c 512 -H 28 -W   28 -m ${1} --forw 2 -s 1 -r 1 | 0.603 | 0.597 | 101% | 0.611 | 0.609 | 100%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 128 -c 512 -H 7 -W 7   -m ${1} --forw 2 -s 1 -r 1 | 0.074 | 0.043 | 173% | 0.075 | 0.102 | 73%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 128 -c 64 -H 112 -W   112 -m ${1} --forw 2 -s 1 -r 1 | 1.037 | 1.059 | 98% | 1.053 | 1.046 | 101%
./bin/MIOpenDriver bnorm -i 10 -t 1 -w 1 -V 0 -n 128 -c 64 -H 56 -W 56   -m ${1} --forw 2 -s 1 -r 1 | 0.307 | 0.293 | 105% | 0.332 | 0.300 | 111%

### CODE REVIEW

**Comment by JehandadKhan** [comment]
@daniellowell Are there any perf numbers ? Would we not have a "Static" version for special cases ?

File: src/kernels/MIOpenBatchNormFwdInferPerAct.cl [hunk 1]
```diff
@@ -35,53 +35,49 @@
...
 
     // PER ACTIVATION
     _FLOAT_PREC mean, variance;
     _FLOAT_PREC invVariance, elemStd, inhat;
     _FLOAT_PREC pvt_scale, pvt_bias;
-    unsigned int adjIndex, inImgIndex, index;
+    unsigned int adjIndex, index;
 
-    int xgid    = get_global_id(0);
+    // int xgid    = get_global_id(0);
```
Comment (muralinr) [inline src/kernels/MIOpenBatchNormFwdInferPerAct.cl:57]: Remove commented code here.

File: src/ocl/batchnormocl.cpp [hunk 2]
```diff
@@ -679,36 +679,36 @@ void BatchNormForwardInference(Handle& handle,
...
+                   bnBias,
+                   epsilon,
+                   n,
+                   in_cstride,
+                   in_nstride,
+                   c);
         }
         else
         {
+            size_t xlocalsize = 1;
```
Comment (muralinr) [inline src/ocl/batchnormocl.cpp:705]: Does these changes impact any performance?

Comment (daniellowell) [reply src/ocl/batchnormocl.cpp:705]: Performance is worse across the board. Attempting to lift it up a bit.

File: src/kernels/MIOpenBatchNormFwdInferPerAct.cl (hunk 1 above)
Comment (daniellowell) [reply src/kernels/MIOpenBatchNormFwdInferPerAct.cl]: done.

File: src/ocl/batchnormocl.cpp (hunk 2 above)
Comment (daniellowell) [reply src/ocl/batchnormocl.cpp:705]: Check the top level comment.

**Comment by daniellowell** [comment]
PR is ready

File: src/kernels/MIOpenBatchNormFwdInferPerAct.cl [hunk 3]
```diff
@@ -35,53 +35,44 @@
...
-            {
-                // per (x-dims) channel load a block of data into LDS
-                index   = MIO_BN_CHW * n + adjIndex;
-                elemStd = (_FLOAT_PREC)(*(in + index)) - mean; // (x_i - mean)
-                inhat   = elemStd * invVariance;
-                out[index] =
-                    (_FLOAT)(mad(pvt_scale, inhat, pvt_bias)); //	y_i = gamma*x_hat + beta
-            }                                                  // end for
-        }                                                      // end if
-    } // end for(img_offset) //image mini_batch is processed
```
Comment (muralinr) [inline src/kernels/MIOpenBatchNormFwdInferPerAct.cl:84]: Does loop unroll improve any performance here?

Comment (daniellowell) [reply src/kernels/MIOpenBatchNormFwdInferPerAct.cl:84]: I can try, but there is not compiler hints, so it would be guessing.

Comment (atamazov) [reply src/kernels/MIOpenBatchNormFwdInferPerAct.cl:84]: FYI Compiler team asked us to report all cases where manual loop unrolling gives some profit. The objective is to get rid of all manual unrolling.

Comment (muralinr) [reply src/kernels/MIOpenBatchNormFwdInferPerAct.cl:84]: yes, We should report such cases.  We should keep loop unrolling if it improves performance for a range of configs..

**Review by muralinr — APPROVED** [review]
